### PR TITLE
SummarizeGradient Layer

### PR DESCRIPTION
**Pull request by Haichao-Zhang**
A gradient summarization layer that can be included in a network for monitoring the gradient information.

This is convenient especially for the tensors that we do not have direct access in the algorithm, e.g. raw input tensors or the tensors involved in preprocessing net.

### CODE REVIEW

File: alf/layers.py [hunk 1]
```diff
@@ -3081,6 +3081,27 @@ def forward(self, input):
...
+            name (str): used to describe the name of the summary, after the
+                tag 'tensor_gradient'.
+        """
+        super().__init__()
+
+        self._name = name
+
+    def forward(self, x):
+        # explicitly turn on gradient calculation in order to summarize gradient
+        x.requires_grad = True
```
Comment (emailweixu) [inline alf/layers.py]: This can change the behavior of training. I suggest  do not summarize if x.requires_grad is false.

Comment (Haichao-Zhang) [reply alf/layers.py]: Changed to only turn-on ``requires_grad`` for leaf nodes, which is a common use case when we want to summarize the gradient w.r.t. inputs.

Comment (Haichao-Zhang) [reply alf/layers.py]: Now changed with a clone of the input as suggested.

**Review by emailweixu — DISMISSED** [review]

**Comment by Haichao-Zhang** [comment]
There is a conflict with a new merge hours ago (#1246). Rebased to pytorch.

**Review by emailweixu — APPROVED** [review]
